### PR TITLE
feat: Image size warning

### PR DESCRIPTION
**Pull request by rhystmills**
## What does this change?

This replicates the image size warning for the Image element, which is present in the existing element in Composer. This warning occurs when the largest asset available has at least one dimension smaller than 460px.

There's quite a bit of code here to allow us remain type-safe while validating the image fields. Is this valuable enough to justify not simply asserting the type using `as`?

## How to test

Run `yarn start` locally and try using a crop with a dimension smaller than 460px.

## Images

| Too Small | Correct Size |
| --- | --- |
|  |  |

## Accessibility

- [ ] [Tested with screen reader](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/558398)
- [X] [Navigable with keyboard](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/39894d)
- [X] [Colour contrast passed](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/92b913)
- [X] [The change doesn't use only colour to convey meaning](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/29032f)

### CODE REVIEW

File: src/plugin/helpers/validation.ts [hunk 1]
```diff
@@ -19,6 +19,82 @@ export const createValidator = (
...
+  prop: Y
+): obj is X & Record<Y, unknown> => {
+  return Object.hasOwnProperty.call(obj, prop);
+};
+
+const isRecord = (value: unknown): value is Record<string, unknown> => {
+  return typeof value === "object" && value !== null;
+};
+
+const isValidImageAsset = (
```
Comment (SHession) [inline src/plugin/helpers/validation.ts:40]: I would recommend moving this logic into a file located within the image element folder, as it is only likely to be useful for that particular element. When the elements are split out, this divide should become clearer.

Comment (rhystmills) [reply src/plugin/helpers/validation.ts:40]: Sounds very sensible - will do so.

File: src/elements/image/ImageElement.tsx [hunk 2]
```diff
@@ -83,6 +84,88 @@ export const createImageFields = (
...
+    hasOwnProperty(maybeImage, "fields") &&
+    isRecord(maybeImage.fields) &&
+    hasOwnProperty(maybeImage.fields, "width") &&
+    hasOwnProperty(maybeImage.fields, "height") &&
+    typeof maybeImage.fields.width === "number" &&
+    typeof maybeImage.fields.height === "number"
+  );
+};
+
+const validateAssets = (maybeAssets: unknown[]) => {
```
Comment (SHession) [inline src/elements/image/ImageElement.tsx:118]: I've come up with a slightly cleaner way of doing this without this function but it isn't much of an improvement. Let me know if you're interested.

**Review by SHession — COMMENTED** [review]
A few things to think about on this. I see your concerns about the code necessary to add typesafety, I think moving it into a separate file should help a lot. Functionally the check behaves well and displays sensibly! Let me know what you think of the comments before I +1

File: src/elements/image/ImageElement.tsx (hunk 2 above)
Comment (rhystmills) [reply src/elements/image/ImageElement.tsx]: Cheers for the review Sam - have implemented your other points, and definitely interested in looking into you suggestion.

Comment (SHession) [reply src/elements/image/ImageElement.tsx]: Had a good discussion about this on a call. We have come up with a solution which is cleaner and more succinct which has now been commited.

**Review by SHession — APPROVED** [review]

**Pull request merged**